### PR TITLE
ReturnnComputePriorJob refactoring

### PR DESCRIPTION
**Pull request by christophmluscher**
* fix hash bug via introducing ReturnnComputePriorJobV2
* some refactoring
* doc

addresses #306

### CODE REVIEW

File: returnn/extract_prior.py [hunk 1]
```diff
@@ -72,52 +78,55 @@ def run(self):
...
-        log_verbosity=3,
-        device="gpu",
-        time_rqmt=4,
-        mem_rqmt=4,
-        cpu_rqmt=2,
-        returnn_python_exe=None,
-        returnn_root=None,
+        log_verbosity: int = 3,
+        device: str = "gpu",
+        time_rqmt: Union[float, int] = 4,
```
Comment (michelwi) [inline returnn/extract_prior.py:104]:
```suggestion
        time_rqmt: float = 4,
```
float implies int

File: returnn/extract_prior.py [hunk 2]
```diff
@@ -72,52 +78,55 @@ def run(self):
...
-            else gs.RETURNN_PYTHON_EXE
-        )
-
-        self.returnn_root = (
-            returnn_root if returnn_root is not None else gs.RETURNN_ROOT
-        )
+        self.returnn_python_exe = returnn_python_exe
+        self.returnn_root = returnn_root
 
         self.returnn_config = ReturnnComputePriorJob.create_returnn_config(**kwargs)
```
Comment (michelwi) [inline returnn/extract_prior.py:131]:
```suggestion
        self.returnn_config = ReturnnComputePriorJobV2.create_returnn_config(**kwargs)
```
I request changes w.r.t. this comment.
If you don't specify V2 here, I think you will still have the hash bug.

**Review by michelwi — CHANGES_REQUESTED** [review]

File: returnn/extract_prior.py (hunk 2 above)
Comment (christophmluscher) [reply returnn/extract_prior.py:131]: > ```suggestion
>         self.returnn_config = ReturnnComputePriorJobV2.create_returnn_config(**kwargs)
> ```
> I request changes w.r.t. this comment.
> If you don't specify V2 here, I think you will still have the hash bug.

Good catch 👍

File: returnn/extract_prior.py [hunk 3]
```diff
@@ -72,54 +78,57 @@ def run(self):
...
-        time_rqmt=4,
-        mem_rqmt=4,
-        cpu_rqmt=2,
-        returnn_python_exe=None,
-        returnn_root=None,
+        log_verbosity: int = 3,
+        device: str = "gpu",
+        time_rqmt: float = 4,
+        mem_rqmt: Union[float, int] = 4,
+        cpu_rqmt: Union[float, int] = 2,
```
Comment (JackTemaki) [inline returnn/extract_prior.py:106]: This is only int...

**Review by JackTemaki — APPROVED** [review]
Only one small comment, the rest "should" be fine...

File: returnn/extract_prior.py (hunk 3 above)
Comment (christophmluscher) [reply returnn/extract_prior.py:106]: true xD

**Review by albertz — APPROVED** [review]

**Review by michelwi — APPROVED** [review]

**Review by JackTemaki — CHANGES_REQUESTED** [review]
same remark as @michelwi

**Comment by christophmluscher** [comment]
> "returnn_config":  ReturnnComputePriorJobV2.create_returnn_config(**kwargs)

done

> also what about ReturnnRasrComputePriorJob?

here a new version would also be necessary, correct?

**Comment by michelwi** [comment]
> > also what about ReturnnRasrComputePriorJob?
> 
> here a new version would also be necessary, correct?

Yes. But probably not much code changes necessay, only inherit from V2 Job.

> new PR?? that would speed things up for this job.. not sure

Ok.

File: returnn/extract_prior.py [hunk 4]
```diff
@@ -72,54 +78,57 @@ def run(self):
...
+        :param returnn_python_exe: path to the RETURNN executable (python binary or launch script)
+        :param returnn_root: path to the RETURNN src folder
         """
         assert isinstance(returnn_config, ReturnnConfig)
         kwargs = locals()
         del kwargs["self"]
 
         self.model_checkpoint = model_checkpoint
 
-        self.returnn_python_exe = (
```
Comment (JackTemaki) [inline returnn/extract_prior.py:112]: If we remove this, we also need to remove the default `None`, otherwise the job will crash.

Comment (michelwi) [reply returnn/extract_prior.py:112]: if we incorporate https://github.com/rwth-i6/i6_core/pull/261 then we could leave default `None` which then falls back to the system python (or some virtual env that is used to run the manager)

File: returnn/extract_prior.py [hunk 5]
```diff
@@ -114,12 +132,11 @@ def __init__(
             if returnn_python_exe is not None
             else gs.RETURNN_PYTHON_EXE
```
Comment (JackTemaki) [inline returnn/extract_prior.py:133]: both `gs.` calls need to be wrapped with `tk.Path`, otherwise later code will not work.

**Review by JackTemaki — CHANGES_REQUESTED** [review]

**Comment by JackTemaki** [comment]
> I thought it was a good feature to _not_ have fallbacks on `gs.RETURNN_ROOT` and `gs.RETURNN_PYTHON_EXE`. Any reason you added them back?

Good point, as this is V2 we could remove those completely.

**Review by michelwi — APPROVED** [review]
I think its good now.

**Review by JackTemaki — APPROVED** [review]

**Pull request merged**